build: add Linux support

TheSunCat opened the pull request:
Hi! I've tested minigin on Linux and made some changes to ensure it builds.

List of changes:
- use `find_package` with modern FindSDL2* modules to find SDL2 and co.
- enable `-Wall -Wextra -Wpedantic -Werror` on non-MSVC compilers, inc. Emscripten
- fix new build errors due to unused member `m_size` and a missing struct member in a `SDL_color` initialization (previously not an issue because `-Werror` was only enabled for MSVC)

TheSunCat committed:
build: add Linux support

TheSunCat committed:
build: match include path with Windows

TheSunCat commented on Minigin/Scene.cpp:
```diff
@@ -1,6 +1,8 @@
 #include "Scene.h"
 #include "GameObject.h"
 
+#include <algorithm>
```
This line was not required in MSVC because of a transitive include; it is required for building on gcc (tested with 13.2.1). It could be useful to add [IWYU](https://include-what-you-use.org/) to the CMake tooling to find more of these.

avadae replied:
Correct

TheSunCat commented on CMakeLists.txt:
```diff
@@ -130,32 +129,54 @@ if (WIN32)
   set(EMSCRIPTEN_LINK_FLAGS "${EMSCRIPTEN_LINK_FLAGS} -sUSE_SDL=2 -sUSE_SDL_IMAGE=2 -sUSE_SDL_TTF=2 -g -sALLOW_MEMORY_GROWTH --preload-file \"${CMAKE_SOURCE_DIR}/Data@/\" -sSDL2_IMAGE_FORMATS=['tga','png']")
 
+  # work around https://github.com/emscripten-core/emscripten/issues/11513
+  set(EMSCRIPTEN_LINK_FLAGS "${EMSCRIPTEN_LINK_FLAGS} -fno-stack-protector")
```
Without this line, the Emscripten build crashes with this log:
```
[100%] Linking CXX executable minigin_web.html
[parse exception: attempted pop from empty stack / beyond block start boundary at 1056407 (at 0:1056407)]
Fatal: error in parsing input
em++: error: '/usr/bin/wasm-emscripten-finalize -g --dyncalls-i64 --pass-arg=legalize-js-interface-exported-helpers --dwarf minigin_web.wasm -o minigin_web.wasm --detect-features' failed (returned 1)
```

Would be good to figure out the root cause (the linked issue says it's fixed...) but for now adding this flag bypasses the crash and allows minigin to run on Firefox.

avadae commented on Minigin/Font.h:
Don't fix it like that, if it is unused then just remove it.

avadae commented on Minigin/Font.h:
```diff
@@ -20,6 +20,6 @@ namespace dae
 	private:
 		_TTF_Font* m_font;
-		unsigned int m_size;
+		[[maybe_unused]] unsigned int m_size;
```
Don't fix it like that, if it is unused then just remove it.

TheSunCat replied:
Fair enough. I figured the field might be there for convenience, but it's better to not have it in the first place unless needed. I've now removed it.

TheSunCat committed:
fix: remove unused m_size member from Font class

TheSunCat commented:
Hi, I think it's not possible to use binaries of SDL for Linux, due to the many different platforms and architectures. It is possible to use FetchContent for it and compile SDL at build-time. It doesn't extend the time too much (~10s in my testing, only for the first build, and only for Linux anyway), so it's an option.
Note that the `Copyright.txt` file for this only applies to the modules, and explicitly allows redistribution. It does not affect the Minigine project itself.

TheSunCat committed:
build: use FetchContent for dependencies on Linux

TheSunCat commented:
I have switched it to use FetchContent for all dependencies. This ups the initial build time (first time takes 80s) but subsequent builds are still almost instant.
This removes the dependency on vendored CMake modules, so I have removed them.
